src/transaction.rs

```diff
@@ -0,0 +1,554 @@
+                .transmit(self, 0xFF)?
+                .status_words();
+
+            if !sw.is_success() {
+                error!("failed selecting application: {:04x}", sw.code());
+                return Err(Error::GenericError);
+            }
+
+            resp
+        } else {
+            // get serial from yk5 and later devices using the f8 command
+            let resp = APDU::new(YKPIV_INS_GET_SERIAL).transmit(self, 0xFF)?;
+
+            if !resp.is_success() {
+                error!(
+                    "failed retrieving serial number: {:04x}",
+                    resp.status_words().code()
+                );
+                return Err(Error::GenericError);
+            }
+
+            resp
+        };
+
+        response.buffer()[..4]
+            .try_into()
+            .map(|serial| Serial::from(u32::from_be_bytes(serial)))
+            .map_err(|_| Error::SizeError)
+    }
+
+    /// Verify device PIN.
+    pub fn verify_pin(&self, pin: &[u8]) -> Result<(), Error> {
+        if pin.len() > CB_PIN_MAX {
+            return Err(Error::SizeError);
+        }
+
+        let response = APDU::new(YKPIV_INS_VERIFY)
+            .params(0x00, 0x80)
+            .data(pin)
+            .transmit(self, 261)?;
+
+        match response.status_words() {
+            StatusWords::Success => Ok(()),
+            StatusWords::AuthBlockedError => Err(Error::WrongPin { tries: 0 }),
+            StatusWords::Other(sw) if sw >> 8 == 0x63 => Err(Error::WrongPin { tries: sw & 0xf }),
+            _ => Err(Error::GenericError),
+        }
+    }
+
+    /// Change the PIN
+    pub fn change_pin(&self, action: i32, current_pin: &[u8], new_pin: &[u8]) -> Result<(), Error> {
+        let mut templ = [0, YKPIV_INS_CHANGE_REFERENCE, 0, 0x80];
+        let mut indata = Zeroizing::new([0u8; 16]);
+
+        if current_pin.len() > 8 || new_pin.len() > 8 {
+            return Err(Error::SizeError);
+        }
+
+        if action == CHREF_ACT_UNBLOCK_PIN {
+            templ[1] = YKPIV_INS_RESET_RETRY;
+        } else if action == CHREF_ACT_CHANGE_PUK {
+            templ[3] = 0x81;
+        }
+
+        unsafe {
+            ptr::copy(current_pin.as_ptr(), indata.as_mut_ptr(), current_pin.len());
+
+            if current_pin.len() < 8 {
+                ptr::write_bytes(
+                    indata.as_mut_ptr().add(current_pin.len()),
+                    0xff,
+                    8 - current_pin.len(),
+                );
+            }
+
+            ptr::copy(
+                new_pin.as_ptr(),
+                indata.as_mut_ptr().offset(8),
+                new_pin.len(),
+            );
+
+            if new_pin.len() < 8 {
+                ptr::write_bytes(
+                    indata.as_mut_ptr().offset(8).add(new_pin.len()),
+                    0xff,
+                    8 - new_pin.len(),
+                );
+            }
+        }
+
+        let status_words = self
+            .transfer_data(&templ, indata.as_ref(), 255)?
```

Comment:
To match earlier invocations:
```suggestion
            .transfer_data(&templ, indata.as_ref(), 0xFF)?
```